1552_ek_certificate_chain.md

```diff
@@ -0,0 +1,155 @@
+
+## Proposal
+
+* keylime_agent, send `ek_ca_chain` to registrar
+* keylime registrar, store `ek_ca_chain` in database
+* keylime tenant, verify `ekcert` against `ek_ca_chain` and `ek_ca_chain` against `tpm_cert_store`
+
+
+### User Stories
+
+#### Story 1
+Users will be able to use keylime with intel firmware TPMs (11th gen core and onwards),
+or any other TPM that stores an EK certificate that is signed by a chain that is stored
+in the TPM.
+
+
+### Risks and Mitigations
+
+#### Registrar/Tenant could be become incompatible with older database
+* Update database to new scheme, only a single key is added to the registar db 'ek_ca_chain'
+
+#### Registrar/Tenant could become incompatible with older Agent
+* Make 'ek_ca_chain' optional
+
+#### Additional memory will be required to store the chain in the database.
+* If the feature can't be used, due to missing certificates in the TPM, the memory footprint will stay around the same.
+
+#### Providing big chains as attack
+* Limit the amount of allowed chain size
+* Use mTLS to only allow verified clients access
+
+
+## Design Details
+First some words from TCG EK documentations:
+
+```txt
+TPM NV MAY contain all certificates of the EK Certificate Chain except the Root CA certificate. The
+EK Certificate Chain MUST be stored as X.509 DER encoded certificates. If the chain consists of
+more than one certificate, or if multiple chains exist, they MUST be stored in NV as a concatenated
+sequence. The TPM manufacturer MAY provision certificate chains using the following list of indices:
+
+0x01c00100 - 0x01c001ff
+
+Revision 2 January 26, 2022 Published Page 15 of 60
+The NV indices MUST be populated starting with index 0x01c00100 through index 0x01c001ff. There
+is no requirement to store the certificates in any particular order whatsoever. If a concatenated
+certificate chain does not fit in a single index, the chain MUST overflow to the next numerically larger
+index in the list of NV Indices. If the storage space in a single index is insufficient to store the entire
+certificate, the certificate MAY overflow into the next numerically larger index in the list of NV Indices.
+```
+
+The feature requires changes in the rust-keylime agent and in the keylime servers registrar and tenant. The certificate chain must and can be only provided by the keylime-agent.
+
+The keylime-agent must do the following:
+* get all available NV-Index-Handles in the defined range (`0X01C00100 - 0X01C001FF`)
+* read the chain from the NV-Index-Handles
+  * Theoretically the certificates can be in any order. This would require an option to make a sorting possible in the agent. As it does not make really sense to store the chain in an unsorted order i would postpone any implementation till this case really exists.
+  * Only a subset of the NV-Index-Handles may be required. This would require an option to filter/set the required handles. I would also wait for implementation till a real case exists for that.
```

Comment:
The rust-tss-esapi is a rust wrapper for the native tss-esapi library, which does not include a function to get the certificate chain. Perhaps it is not such a bad idea to have a custom implementation. The guidelines in the "TCG EK Credential Profile" provide manufacturers with considerable freedom in how to store certificates in the TPM. Additionally, there is the possibility of storing multiple certificate chains. In my opinion, the role of the Keylime agent is to read, filter, and sort the certificates. It should provide the Keylime registrar with exactly **one** verifiable chain at any given time. The server should not be responsible for managing multiple chains for clients or determining which chain can be verified. The location of the chain and how it should be processed (filtered/sorted) should be configurable through the agent's settings. This approach leads to the majority of the work (filtering/sorting) being handled by the agent, making the call to rust-tss-esapi equivalent to an nvread operation.

Reply:
Thanks for the explanation. I can see that it then makes sense to implement in agent. If it is generic enough we can upstream it to the abstraction layer of the rust-tss-espai crate, which provides more high level common functions.

On another note we have our monthly meeting today: https://github.com/keylime/meetings/issues/82. Would be great if you have the time to join.

Reply:
Sure, thank you. I will join.

Reply:
I'm not sure if I agree completely with adding the possibility for the agent to be configured on how to sort and filter certificates. For me, the agent simply has to find the chain that signed the EK certificate and provide it, no need for the other chains (why they are there in the first place?). Probably being able to configure the range of NV indexes to read is sufficient, but the default will be to read all of them anyway (i.e. in case the range is not configured).

I agree with @THS-on that getting the certificate chain looks generic enough to be implemented on `rust-tss-esapi` crate. It is definitely not specific to Keylime.